go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+	defer func() {
+		klog.V(5).Infof("Finished reconciling health check for node %s: %v", nodeName, time.Since(start))
+	}()
+
+	node, err := c.nodeInformer.Lister().Get(nodeName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if node == nil {
+		// node deleted, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node deleted", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	names := []string{}
+	if !c.hasInterestedConsumers(nodeName, &names) {
+		// no interest, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: no interested consumers out of %q", nodeName, names)
+			// in case someone just registered
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeSubnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		return fmt.Errorf("failed to parse node %s subnets annotation %v", node.Name, err)
+	}
+
+	mgmtIPs := make([]net.IP, len(nodeSubnets))
+	for i, subnet := range nodeSubnets {
+		mgmtIPs[i] = util.GetNodeManagementIfAddr(subnet).IP
+	}
+
+	update := true
+	nodeInfo, probed := c.isProbedWithNodeInfo(nodeName)
+	if probed {
+		less := func(a, b net.IP) bool { return a.String() < b.String() }
+		update = !cmp.Equal(nodeInfo.ips, mgmtIPs, cmpopts.SortSlices(less))
+	}
+
+	if !update {
+		// already probing and no config update, bail out
+		return nil
+	}
+
+	if len(mgmtIPs) == 0 {
+		// no proper config, stop probing and assume unreachable out of
+		// precaution
+		updated := c.cancelProbeWithState(nodeName, UNREACHABLE)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node management IPs unknown", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeInfo.ips = mgmtIPs
+
+	// add/update probing task
+	klog.Infof("Starting or updating health state probing for node %s on IPs %s", nodeName, util.StringSlice(mgmtIPs))
+	c.startProbeWithNodeInfo(nodeName, nodeInfo)
+
+	return nil
+}
+
+func (c *controller) informConsumers(node string) {
+	c.consumerLock.RLock()
+	defer c.consumerLock.RUnlock()
+	// note we don't query the interest of the consumers as that might be racey,
+	// it is up to the consumers to appropriately filter out uninteresting
+	// events
+	for _, consumer := range c.consumers {
+		consumer.HealthStateChanged(node)
+	}
+}
+
+func (c *controller) hasInterestedConsumers(node string, names *[]string) bool {
+	c.consumerLock.RLock()
+	defer c.consumerLock.RUnlock()
+	for _, consumer := range c.consumers {
+		if names != nil {
+			*names = append(*names, consumer.Name())
+		}
+		if consumer.MonitorHealthState(node) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) isProbedWithNodeInfo(node string) (nodeInfo, bool) {
```

Comment:
why isn't this just called "getNodeInfo"?

Reply:
The presence of node info actually means that the node is being probed. I thought this name was better in that  context.

Reply:
imho I think it adds confusion. If the nodeInfo is contained in the healtheck controller, its safe to assume it is being healthchecked. I would just change it to getNodeInfo and modify the comment in the code:

```
	// nodeInfo holds information about each node that will be healthchecked
	nodeInfo  map[string]nodeInfo
        // nodeState holds the latest status of the node's health/reachability
	nodeState map[string]nodeState
```